strawberryfields/circuitspecs/chip2.py

```diff
@@ -0,0 +1,260 @@
+        Rgate({phase_11}) | [1]
+        BSgate(pi/4, pi/2) | [1, 2]
+
+        # duplicate the interferometer for the idler modes (the higher ones in frequency)
+        # MZI_0
+        Rgate({phase_0}) | [4]
+        BSgate(pi/4, pi/2) | [4, 5]
+        Rgate({phase_1}) | [4]
+        BSgate(pi/4, pi/2) | [4, 5]
+        # MZI_1
+        Rgate({phase_2}) | [6]
+        BSgate(pi/4, pi/2) | [6, 7]
+        Rgate({phase_3}) | [6]
+        BSgate(pi/4, pi/2) | [6, 7]
+        # MZI_2
+        Rgate({phase_4}) | [5]
+        BSgate(pi/4, pi/2) | [5, 6]
+        Rgate({phase_5}) | [5]
+        BSgate(pi/4, pi/2) | [5, 6]
+        # MZI_3
+        Rgate({phase_6}) | [4]
+        BSgate(pi/4, pi/2) | [4, 5]
+        Rgate({phase_7}) | [4]
+        BSgate(pi/4, pi/2) | [4, 5]
+        # MZI_4
+        Rgate({phase_8}) | [6]
+        BSgate(pi/4, pi/2) | [6, 7]
+        Rgate({phase_9}) | [6]
+        BSgate(pi/4, pi/2) | [6, 7]
+        # MZI_5
+        Rgate({phase_10}) | [5]
+        BSgate(pi/4, pi/2) | [5, 6]
+        Rgate({phase_11}) | [5]
+        BSgate(pi/4, pi/2) | [5, 6]
+
+        # add final dummy phases to allow mapping any unitary to this template (these do not
+        # affect the photon number measurement)
+        Rgate({final_phase_0}) | [0]
+        Rgate({final_phase_1}) | [1]
+        Rgate({final_phase_2}) | [2]
+        Rgate({final_phase_3}) | [3]
+        Rgate({final_phase_4}) | [4]
+        Rgate({final_phase_5}) | [5]
+        Rgate({final_phase_6}) | [6]
+        Rgate({final_phase_7}) | [7]
+
+        # measurement in Fock basis
+        MeasureFock() | [0, 1, 2, 3, 4, 5, 6, 7]
+        """
+    )
+
+    def compile(self, seq, registers):
+        """Try to arrange a quantum circuit into a form suitable for Chip0.
+
+        Args:
+            seq (Sequence[Command]): quantum circuit to modify
+            registers (Sequence[RegRefs]): quantum registers
+        Returns:
+            List[Command]: modified circuit
+        Raises:
+            CircuitError: the circuit does not correspond to Chip0
+        """
+        # pylint: disable=too-many-statements,too-many-branches
+        # First, check if provided sequence matches the circuit template.
+        # This will avoid superfluous compilation if the user is using the
+        # template directly.
+        try:
+            seq = super().compile(seq, registers)
+        except CircuitError:
+            # failed topology check. Continue to more general
+            # compilation below.
+            pass
+        else:
+            return seq
+
+        # first do general GBS compilation to make sure
+        # Fock measurements are correct
+        # ---------------------------------------------
+        seq = GBSSpecs().compile(seq, registers)
+        A, B, C = group_operations(seq, lambda x: isinstance(x, ops.MeasureFock))
+
+        if len(B[0].reg) != self.modes:
+            raise CircuitError("All modes must be measured.")
```

Comment:
I can't quite remember, but I think the reason we enforce all programs ending with `MeasureFock` on all modes is so that hardware simulation output matches software simulation output. The hardware will always return samples from all modes, whereas simulators currently return `None` if the mode is not explicitly measured.

This is to avoid the same compiled program returning different values on different backends.

Reply:
But we can easily put logic in HW jobs pipeline to not return results for unwanted measured modes, can't we? Can save on the size of results that has to be communicated. 

Anyway, we shouldn't tie too closely to what the simulators return, since we have a long-open ticket to fix up and unify that

Reply:
Good point. I suppose the logic for template matching would need to be modified to mark lines as optional? :thinking:

Reply:
the measurement operator will need to be non-optional, and some (logically consistent) set of modes must be provided. But we should probably allow for the modes to be a legal a subset (not important for current time, but something to plan going forward)

Reply:
I think this is looking complicated enough to be its own issue/PR? It also applies to other circuitspecs not touched in this PR.